exec/java-exec/src/test/java/org/apache/drill/exec/server/TestSpnegoAuthentication.java

```diff
@@ -0,0 +1,597 @@
+    final String token = Subject.doAs(clientSubject, new PrivilegedExceptionAction<String>() {
+      @Override
+      public String run() throws Exception {
+
+        final GSSManager gssManager = GSSManager.getInstance();
+        GSSContext gssContext = null;
+        try {
+          final Oid oid = GSSUtil.GSS_SPNEGO_MECH_OID;
+          final GSSName serviceName = gssManager.createName(spnegoHelper.SERVER_PRINCIPAL, GSSName.NT_USER_NAME, oid);
+
+          gssContext = gssManager.createContext(serviceName, oid, null, GSSContext.DEFAULT_LIFETIME);
+          gssContext.requestCredDeleg(true);
+          gssContext.requestMutualAuth(true);
+
+          byte[] outToken = new byte[0];
+          outToken = gssContext.initSecContext(outToken, 0, outToken.length);
+          return Base64.encodeBase64String(outToken);
+
+        } finally {
+          if (gssContext != null) {
+            gssContext.dispose();
+          }
+        }
+      }
+    });
+
+    // Create a DrillbitContext with service principal and keytab for DrillSpnegoLoginService
+    final DrillConfig newConfig = new DrillConfig(DrillConfig.create()
+        .withValue(ExecConstants.HTTP_AUTHENTICATION_MECHANISMS,
+            ConfigValueFactory.fromIterable(Lists.newArrayList("spnego")))
+        .withValue(ExecConstants.HTTP_SPNEGO_PRINCIPAL,
+            ConfigValueFactory.fromAnyRef(spnegoHelper.SERVER_PRINCIPAL))
+        .withValue(ExecConstants.HTTP_SPNEGO_KEYTAB,
+            ConfigValueFactory.fromAnyRef(spnegoHelper.serverKeytab.toString())),
+        false);
+
+
+    final SystemOptionManager optionManager = Mockito.mock(SystemOptionManager.class);
+    Mockito.when(optionManager.getOption(ExecConstants.ADMIN_USERS_VALIDATOR))
+        .thenReturn(ExecConstants.ADMIN_USERS_VALIDATOR.DEFAULT_ADMIN_USERS);
+    Mockito.when(optionManager.getOption(ExecConstants.ADMIN_USER_GROUPS_VALIDATOR))
+        .thenReturn(ExecConstants.ADMIN_USER_GROUPS_VALIDATOR.DEFAULT_ADMIN_USER_GROUPS);
+
+    final DrillbitContext drillbitContext = Mockito.mock(DrillbitContext.class);
+    Mockito.when(drillbitContext.getConfig()).thenReturn(newConfig);
+    Mockito.when(drillbitContext.getOptionManager()).thenReturn(optionManager);
+
+    final DrillSpnegoLoginService loginService = new DrillSpnegoLoginService(drillbitContext);
+
+    // Authenticate the client using its SPNEGO token
+    final UserIdentity user = loginService.login(null, token);
+
+    // Validate the UserIdentity of authenticated client
+    assertTrue(user != null);
+    assertTrue(user.getUserPrincipal().getName().equals(spnegoHelper.CLIENT_PRINCIPAL));
+    assertTrue(user.isUserInRole("authenticated", null));
+  }
+
+  /**
+   * Validate {@link DrillSpnegoAuthenticator} with request of different nature and from authenticated and
+   * unauthenticated session.
+   * @throws Exception
+   */
+  @Test
+  public void testDrillSpnegoAuthenticator() throws Exception {
```

Comment:
The same here, test can be factored out in the separate class dedicated `DrillSpnegoAuthenticator` testing and split into several.